[docker] Add Docker client role

imrejonk opened the pull request:
Alright! It was a bit more work than anticipated, but the lion's share of the new Docker client role is done. The new role manages Docker CLI installation (from upstream, there is no docker.io-cli package in Debian) and configuration, and also manages Docker components by itself. Docker servers provisioned with debops.docker_server can soon be managed from the host itself or from remote hosts via the Docker HTTPS API, thanks to debops.pki. The components that can be managed currently are Docker networks, registry accounts and containers (as implemented by the corresponding Ansible modules).

### What needs to happen before this role can be merged?

- [x] I need feedback on the progress so far. The more, the better!
- [x] The role needs to be improved after receiving feedback.
- [x] The role needs updating for and testing with Ansible 2.9.
- [x] Tests have to be configured. I haven't done this for this project before, so some guidance might be needed.

Closes #870.

imrejonk committed:
[debops.docker_client] Add new role

imrejonk committed:
Update Ansible modules RST defs

imrejonk committed:
[debops.docker] Finish Docker client role

drybjed commented on CHANGELOG.rst:
```diff
@@ -36,6 +36,9 @@ Added
 
+- [debops.docker] New role for Docker client management. The old debops.docker
+  role has been renamed to :ref:`debops.docker_server`.
+
```
Usually new DebOps roles have their own subsection in the "Added" section to combine them in one place. You could add it if it isn't in the Changelog yet, see previous releases for how it was written.

drybjed commented on ansible/roles/debops.docker/defaults/main.yml:
```diff
@@ -0,0 +1,391 @@
+---
+# .. vim: foldmarker=[[[,]]]:foldmethod=marker
+
+# debops.docker_client default variables
```
This needs to be renamed to `debops.docker`.

drybjed commented on ansible/roles/debops.docker/defaults/main.yml:
```diff
@@ -0,0 +1,391 @@
+
+
+# APT packages [[[
+# ------------------------------
```
The underline should be flush with the section title. It's not a bug in the Sphinx parser, but it looks better in the source code.

drybjed commented on ansible/roles/debops.docker/defaults/main.yml:
```diff
@@ -0,0 +1,391 @@
+# - on-failure
+# - always
+# - unless-stopped
+docker__restart_policy: 'no'
```
DebOps is aimed at production environments. I'm not sure what is the convention for Docker in production, shouldn't failed containers always restart automatically? Perhaps a default of `unless-stopped` would work better in this case?

imrejonk replied:
Well we only use `unless-stopped` in production, but `no` is the Docker default (which I just copied). I'll change it to `unless-stopped`.

drybjed commented on ansible/roles/debops.docker/tasks/main.yml:
```diff
@@ -0,0 +1,55 @@
+---
```
Usually there's an empty space here, but it's not a YAML convention or a requirement as far as I know.

drybjed commented on ansible/roles/debops.docker/tasks/main.yml:
```diff
@@ -0,0 +1,55 @@
+    path: '/root/.docker'
+    state: 'directory'
+    owner: 'root'
+    group: 'root'
```
The role is executed as `root` already, you should drop the `owner` and `group` in such cases. Use them only if you want to change the ownership to a specific account/group.

drybjed commented on ansible/roles/debops.docker/tasks/manage_components.yml:
```diff
@@ -0,0 +1,148 @@
+    docker__register_crt_path: '{{ docker__pki_path + "/" + docker__pki_realm
+                                   + "/" + docker__pki_crt }}'
+    docker__register_key_path: '{{ docker__pki_path + "/" + docker__pki_realm
+                                   + "/" + docker__pki_key }}'
```
These variables should be defined in the `defaults/main.yml` file instead, you won't need a separate task to set them up.

imrejonk replied:
Should I add new `docker__pki_*_path` variables or just replace the `docker__pki_*` ones?

imrejonk replied:
I went with just replacing the docker__pki_* variables, that's probably the cleanest solution.

drybjed commented on docs/ansible/roles/debops.docker/defaults-detailed.rst:
```diff
@@ -0,0 +1,196 @@
+   :local:
+   :depth: 1
+
+.. _docker_ref_networks:
```
You should add a link near the variables in the `defaults/main.yml` file to their respective documentation sections, it will be easier to reach on the documentation page.

drybjed commented on docs/ansible/roles/debops.docker/defaults-detailed.rst:
```diff
@@ -0,0 +1,196 @@
+  the Docker host. If ``absent``, the network will be removed.
+
+Examples
+~~~~~~~~
```
The "Examples" section should be put before the "Syntax" section, that way you get a quick overview of how the configuration is written, and you can look up the specific syntax later on.

drybjed reviewed (changes requested):
Sorry for a late review. Very nice work on the PR! There are a few things I'd like you to change or explain, see the comments.

I wonder what's the default use case for `debops.docker`, do you use it to control each Docker host and set up the containers during deployment, or do you have a dedicated Docker client VM/host from which you control remote Docker instances? On a brief look in the [Docker CLI documentation](https://docs.docker.com/engine/reference/commandline/cli/) I don't see any nice way to define multiple hosts, I guess that's done using the environment variables in scripts?

imrejonk commented:
Hey @drybjed, thanks for the feedback! Hope you had a great holiday :)

The Docker host should be managed with `debops.docker_server`. This role takes care of installing the Docker CLI and Docker Compose (on the Docker host or any other DebOps-managed machine), and allows for defining Docker components like networks, registry accounts and containers in Ansible variables. It works just like `debops.mariadb`.

It is possible to run the role against another machine than your Docker host. In that case the docker module tasks will be executed over HTTPS (thanks to `debops.pki`), connecting to the host in `docker__docker_host`. The Docker host and TLS details will also be added to the respective environment variables in /etc/environment for CLI usage. The Docker host and TLS details can be overridden in the `docker__*_networks`, `docker__*_registry_accounts` and `docker__*_containers` variables.

imrejonk committed:
[debops.docker] Apply feedback from drybjed

drybjed commented:
@imrejonk The Travis-CI test seems to be stuck for some reason. Hopefully it will be fixed when you resolve the conflicts.

drybjed commented on ansible/roles/debops.docker/tasks/main.yml:
```diff
@@ -0,0 +1,54 @@
+    repo: '{{ docker__upstream_repository }}'
+    mode: '0644'
+    state: 'present'
+    update_cache: True
```
Can you drop the abolve two tasks and switch to using the `debops.keyring` role? The `debops.docker_server` has a good example on how to do this, created via https://github.com/debops/debops/commit/85c1ddf7270f50e860a2049431a8040547f72e04 commit.

imrejonk committed:
[debops.docker] APT keys via dependency

imrejonk committed:
[debops.docker] Add playbook to virt/all.yml

imrejonk commented:
Nice work on the `debops.keyring` role :)
I saw your comment on #870, I'll see if I can modify the `debops.docker` role so that it can be used as a dependency by other roles.

imrejonk committed:
[debops.docker] Fix /root/.docker/ PKI symlinks

imrejonk committed:
[debops.docker] Allow use as dependency

imrejonk committed:
[debops.docker] Use separate variable for FQDN

The Docker host FQDN is now stored in docker__docker_host_fqdn instead
of in two other variables.

imrejonk commented:
@drybjed do you know why Travis is complaining? Something about a missing global.rst file, but I didn't touch that file in my last commits. `make docs` is working fine here.

drybjed commented:
@imrejonk If you rebased recently, a few hours ago I modified the documentation generation to support multiple `defaults/main/*.yml` files in the roles. You have to add `../` to the include line in `defaults/main.yml` because the generated `.rst` file has been moved to a subdirectory.

imrejonk committed:
[debops.docker] Update documentation includes

This is required for multiple 'defaults/' files support.

imrejonk commented:
Ah, I see. I also had to update the toctree section in index.rst, Travis is happy once again :+1:

imrejonk committed:
[debops.docker] Apply APT and PKI bugfixes

- Do not install docker-ce-cli when Docker server is already installed
- Remove conflicting DOCKER_TLS envvar (implied by DOCKER_TLS_VERIFY)

imrejonk committed:
[debops.docker_server] Update local facts

The docker_server fact now checks if the `docker-init` command is
available to determine if Docker server is installed. This is more
reliable as the CLI package, docker-ce-cli, also provides a `docker`
command.

drybjed commented:
With pull requests, it might be better if you `git rebase` your branch over the `master` branch and force-push the changes into the PR - the git commits are easier to see in the graph that way, vs. merging the changes from the `master` branch into which you will be merged anyway.

Ref: https://stackoverflow.com/questions/804115/when-do-you-use-git-rebase-instead-of-git-merge

imrejonk commented:
You're right. I didn't rebase these commits because I ended up with lots of conflicts and then decided to do a good 'ol merge. I'll try again next time...

imrejonk commented:
Ah, much better :)

imrejonk committed:
[docs] Explain what debops.docker_server does

imrejonk committed:
[docker] Remove 'debops.' prefix

imrejonk committed:
[docker] Fix playbook syntax and REUSE compliance

imrejonk committed:
[docker] Update role for Ansible 2.9

imrejonk committed:
[docker] Apply minor documentation and task fixes

imrejonk committed:
[docker] Add CI config for Docker role testing

imrejonk committed:
[docker] Add meta/main.yml

imrejonk commented:
@drybjed @ypid I think the new role is done. Would you care for another review?

I'm thinking about removing the Docker client functionality from debops.docker_server in a separate PR, there are way too many changes in this one already.

imrejonk commented:
@drybjed I see that this PR is still open. I can (quite easily) update it for Ansible 2.10, or I can close it and be done with it.

I should note that I don't personally use this role anymore and so will not be actively maintaining it in the future.
